### DISCUSSION TITLE
Mocking Issue With Transaction

### DISCUSSION THREAD

**Discussion by TheoCasner** [prisma/prisma#14094] [Q&A]
Hello everyone,

I am currently using Prisma with NestJS, and I have an issue when willing to test my transactions.
I have a service which starts a transaction as following and I pass the TransactionClient created inside the transaction function to my repositories to be able to use the same client on all of the calls or transaction wouldn't work.

But I have no idea how I can test this transaction in unit tests because I would like to mock the calls to this prismaService (extends prisma client).
How can I mock `this.prismaService.$transaction(async (transaction) => { ...` ?

My Service below:

```ts
@Injectable()
export class AssociationsService {
  constructor(
    private readonly prismaService: PrismaService,
    private readonly associationRepository: AssociationRepository,
    private readonly roleRepository: RoleRepository,
    private readonly associationsMemberRepository: AssociationsMemberRepository
  ) {}

  public async create(createAssociationDto: CreateAssociationDto): Promise<AssociationDto> {
    return await this.prismaService.$transaction(async (transaction) => {
      try {
        const createdAsso = await this.associationRepository.create(
          {
            name: createAssociationDto.name,
            description: createAssociationDto.description,
          },
          transaction
        );
        const { id } = await this.roleRepository.createRolePresident(createdAsso.id, transaction);
        await this.associationsMemberRepository.linkUserToRole(
          createdAsso.id,
          createAssociationDto.presidentId,
          id,
          transaction
        );
        return createdAsso;
      } catch (error) {
        // handle error
      }
    });
  }
}
```

One repository example:
```ts
@Injectable()
export class AssociationRepository {
  constructor(private prisma: PrismaService) {}

  public async create(
    createAssociation: any,
    currentTransaction: Prisma.TransactionClient = null
  ): Promise<Association> {
    const client = currentTransaction ? currentTransaction : this.prisma;
    return client.association.create({ data: createAssociation });
  }
}
```

Do you have any idea how to achieve this ? Because I do not want to make calls to database directly in my service but use repositories. And in my testing strategy for services I mock the repositories.

Best regards,
Théo

**Comment by austincrim** [comment]
Hey @TheSalfy 👋,

What is your unit testing setup? Are you following [this guide](https://www.prisma.io/docs/guides/testing/unit-testing)?

**Reply by TheoCasner** [reply]
Hello @austincrim !
I am using mocks, I basically mock everything that connects to database, therefore, I wanna test transactions like said above but I cannot find a correct setup to do so. 
Because I mock return values from functions, but I cannot/or do not know how to mock the transactionClient given in the lambda where I generate transactions.

**Reply by pantharshit00** [reply]
Hello @TheoCasner, our apologies for the delay in getting back to you! Due to the large number of questions in our community channels, unfortunately this slipped through the cracks. It's very important for us to address your question, so let us know in a comment if this is still an open discussion. 


If it is a single test, following might work depending on a usecase:
```ts
jest.spyOn(prisma, '$transaction').mockImplementation((callback) => {
  return callback(prisma)
})
```
and then mock methods on regular prisma objects. Caveats:
1. It won’t be running in actual transaction, so if anything goes wrong no changes will be rolled back. But I assume the goal is to avoid actual DB anyway, otherwise, mocking would’n have came up
2. Since we are passing actual prisma instance to the tx callback, it can do more than an actual transaction object will be able to: bunch of methods are not available on real tx proxies. TS will be able to prevent it, but there will be no corresponding runtime check

**Reply by pantharshit00** [reply]
Hello @TheoCasner,

I was just going through past discussion and making sure everything is addressed. Can you please confirm here that this issue was addressed?

**Reply by TheoCasner** [reply]
Hello ! 
Issue is addressed. Thank you

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
